python/ray/train/function_trainer.py

```diff
@@ -0,0 +1,121 @@
+import inspect
+from typing import Any, Optional, Callable, Type
+
+from ray.air.checkpoint import Checkpoint
+from ray.air.config import RunConfig, ScalingConfig
+from ray.train.trainer import BaseTrainer
+from ray.tune.trainable import Trainable, wrap_function
+from ray.tune.utils import detect_checkpoint_function, detect_config_single
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class FunctionTrainer(BaseTrainer):
+    """Trainer for generic (non-distributed) training functions.
+
+    This trainer can be used to wrap a generic function trainable. In contrast
+    to e.g. the DataParallelTrainer, this trainer will only start up one worker
+    (the trainer). Starting up potential distributed workers is then up
+    to the custom training function.
+
+    The trainer accepts a ``train_fn``, which can be any Ray Tune-compatible
+    function trainable.
+
+    If a scaling config is given, resources will be requested for all workers.
+    Starting potential distributed workers within the current placement group
+    is up to the user.
+
+    If a checkpoint to resume from is given, it will be passed as the
+    ``checkpoint_dir`` argument of the training function, if supported.
+    Alternatively, ``ray.air.session.get_checkpoint()`` can be used.
+
+    This trainer does not handle datasets and preprocessors. If this is needed,
+    consider implementing a DataParallelTrainer with only one worker instead.
+
+    Args:
+        train_fn: Callable training function taking a ``config`` argument.
+        scaling_config: Configuration for how to scale data parallel training.
+        run_config: Configuration for the execution of the training run.
+        resume_from_checkpoint: A checkpoint to resume training from.
+
+    Example:
+
+        .. code-block:: python
+
+            from ray.air.config import ScalingConfig
+            from ray.train.function_trainer import FunctionTrainer
+
+            def train_fn(config):
+                # ...
+                return {"metric": 5}
+
+            trainer = FunctionTrainer(
+                train_fn,
+                scaling_config=ScalingConfig(
+                    trainer_resources={"CPU": 4}
+                )
+            )
+            trainer.fit()
+
+    """
+
+    _scaling_config_allowed_keys = BaseTrainer._scaling_config_allowed_keys + [
+        "num_workers",
```

Comment:
Why do we allow num_workers?

Reply:
The original purpose of this trainer was to add a way to add `resources_per_trial` functionality to the Tuner API for functional parity with tune.run() ([see this doc](https://docs.google.com/document/d/1dbRvmSYRB4E0BZSETPoj4g33eq5P3kwfVVp5FFQsSuI/edit#heading=h.2z2oq296adj)). The three options we discussed were

1. Add as a parameter to RunConfig
2. Introduce `tune.with_resources()`
3. Introduce a function trainer

A general feedback was to use the ScalingConfig rather than a resource definition / placement group factory so users don't have to learn different concepts. To achieve this and still support requesting multiple bundles (as users can to with resources_per_trial), we keep the default scaling config including num_workers, but we just don't start these workers ourselves.

An alternative to this would be to introduce a different key to request additional bundles, or to stick with placement group factories. In that case, we should probably stick with 1 or 2, though.